bin/pytorch_inference/Main.cc

```diff
@@ -0,0 +1,183 @@
+    TFloatVec tokens;
+    std::uint32_t numTokens;
+    if (readUInt32(inputStream, numTokens) == false) {
+        LOG_ERROR(<< "Error reading the number of tokens");
+        return boost::none;
+    }
+
+    for (uint32_t i = 0; i < numTokens; ++i) {
+        std::uint32_t token;
+        if (readUInt32(inputStream, token) == false) {
+            LOG_ERROR(<< "Error reading token");
+            return boost::none;
+        }
+        tokens.push_back(token);
+    }
+
+    return tokens;
+}
+
+void writePrediction(torch::Tensor& prediction, std::ostream& outputStream) {
+    rapidjson::OStreamWrapper writeStream(outputStream);
+    ml::core::CRapidJsonLineWriter<rapidjson::OStreamWrapper> jsonWriter(writeStream);
+    jsonWriter.StartObject();
+    jsonWriter.Key("inference");
+    jsonWriter.StartArray();
+    auto arr = prediction.accessor<std::int64_t, 2>();
+    for (int i = 0; i < arr.size(1); i++) {
+        jsonWriter.Int64(arr[0][i]);
+    }
+    jsonWriter.EndArray();
+    jsonWriter.EndObject();
+}
+
+int main(int argc, char** argv) {
+    // command line options
+    std::string modelId;
+    std::string inputFileName;
+    bool isInputFileNamedPipe{false};
+    std::string outputFileName;
+    bool isOutputFileNamedPipe{false};
+    std::string restoreFileName;
+    bool isRestoreFileNamedPipe{false};
+    std::string persistFileName;
+    ml::core_t::TTime namedPipeConnectTimeout{
+        ml::core::CBlockingCallCancellingTimer::DEFAULT_TIMEOUT_SECONDS};
+
+    if (ml::torch::CCmdLineParser::parse(
+            argc, argv, modelId, namedPipeConnectTimeout, inputFileName,
+            isInputFileNamedPipe, outputFileName, isOutputFileNamedPipe,
+            restoreFileName, isRestoreFileNamedPipe) == false) {
+        return EXIT_FAILURE;
+    }
+
+    ml::core::CBlockingCallCancellingTimer cancellerThread{
+        ml::core::CThread::currentThreadId(), std::chrono::seconds{namedPipeConnectTimeout}};
+
+    // Construct the IO manager before reconfiguring the logger, as it performs
+    // std::ios actions that only work before first use
+    const std::string EMPTY;
+    ml::api::CIoManager ioMgr{cancellerThread,
+                              inputFileName,
+                              isInputFileNamedPipe,
+                              outputFileName,
+                              isOutputFileNamedPipe,
+                              restoreFileName,
+                              isRestoreFileNamedPipe,
+                              EMPTY,
+                              false};
+
+    if (cancellerThread.start() == false) {
+        // This log message will probably never been seen as it will go to the
+        // real stderr of this process rather than the log pipe...
+        LOG_FATAL(<< "Could not start blocking call canceller thread");
+        return EXIT_FAILURE;
+    }
+    if (ml::core::CLogger::instance().reconfigure(
+            EMPTY, EMPTY, cancellerThread.hasCancelledBlockingCall()) == false) {
```

Comment:
I added a log pipe argument to the program and now log to the pipe. `evaluate.py` reads the log pipe output but in the main thread for now. 

I also removed the `--xxxIsPipe` arguments as pipes are the only option right now

Reply:
> I added a log pipe argument to the program and now log to the pipe. evaluate.py reads the log pipe output but in the main thread for now.

Thanks.

> I also removed the `--xxxIsPipe` arguments as pipes are the only option right now

One thing to note about this is that it will add complexity in the production Java code in order to simplify a dev Python script - see https://github.com/elastic/elasticsearch/blob/d613a1a1d12890825b982c1a7a4fd6e17f20b325/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/ProcessPipes.java#L110-L133

Theoretically on-disk files could have been used for the inputs if someone put the appropriate binary data in them.  We sometimes use this technique for running autodetect through valgrind.

I would keep the arguments consistent between all the processes that connect to Java using named pipes.

Reply:
I took out the `--xxxIsPipe` arguments because that is not the intended usage but I can see the sense in keeping the arguments consistent with the other apps.  

It also made me realise the C++ can be launched by the python script making testing a lot easier as you just have to run the python now.